numpy/distutils/tests/test_system_info.py

```diff
@@ -285,3 +285,34 @@ def test_overrides(self):
         finally:
             os.chdir(previousDir)
         
+
+def test_distutils_parse_env_order():
+    from numpy.distutils.system_info import _parse_env_order
+    env = 'NPY_TESTS_DISTUTILS_PARSE_ENV_ORDER'
+
+    base_order = list('abcdef')
+
+    os.environ[env] = 'b,i,e,f'
+    order, reject, unknown = _parse_env_order(base_order, env)
+    assert len(order) == 3
+    assert order == list('bef')
+    assert len(reject) == len(base_order) - 3
+    assert reject == list('acd')
+    assert len(unknown) == 1
+
+    for prefix in '^!':
+        os.environ[env] = f'{prefix}b,i,e'
+        order, reject, unknown = _parse_env_order(base_order, env)
+        assert len(order) == 4
+        assert order == list('acdf')
+        assert len(reject) == len(base_order) - 4
+        assert reject == list('be')
+        assert len(unknown) == 1
+
+    with pytest.raises(ValueError):
+        os.environ[env] = 'b,^e,i'
+        _parse_env_order(base_order, env)
+
+    with pytest.raises(ValueError):
+        os.environ[env] = '!b,^e,i'
```

Comment:
This leaves the environment in a bad state, which may affect other tests run after it. I don't care much for very elaborate cleanup, but unsetting the env var at the end of this test would be nice.